rdwatch/core/tasks/__init__.py

```diff
@@ -762,3 +767,118 @@ def download_sam_model_if_not_exists(**kwargs):
+    if len(region_models) > 1:
+        raise ValueError('Too many regions provided in the zip file')
+
+    region_model = region_models[0]
+    region_feature = region_model.region_feature
+    region_id = region_feature.properties.region_id
+    region_originator = region_feature.properties.originator
+
+    # validate site models against the single region model
+    for model in site_models:
+        # if there is an override region ID,
+        # then set the region ID on the site model
+        if model_run_upload.region:
+            model.site_feature.properties.region_id = model_run_upload.region
+        elif model.site_feature.properties.region_id != region_id:
+            raise ValueError("Site model doesn't reference the region model")
+
+        # if there is an override performer,
+        # then set the performer on the site model
+        if model_run_upload.performer:
+            model.site_feature.properties.originator = model_run_upload.performer
+        elif model.site_feature.properties.originator != region_originator:
+            raise ValueError(
+                "Site model's originator differs from the region model\'s originator"
+            )
+
+    # apply overrides to the region model
+    if model_run_upload.region:
+        region_feature.properties.region_id = model_run_upload.region
+    if model_run_upload.performer:
+        region_feature.properties.originator = model_run_upload.performer
+        for site_summary_feature in region_model.site_summary_features:
+            site_summary_feature.properties.originator = model_run_upload.performer
+
+    with transaction.atomic():
+        # create a new ModelRun
+        region, _ = get_or_create_region(
+            region_feature.properties.region_id,
+            region_feature.parsed_geometry,
+        )
+
+        performer_shortcode = region_feature.properties.originator
+        performer, created = Performer.objects.get_or_create(
+            short_code=performer_shortcode.upper(),
+        )
+        if created:
+            performer.team_name = performer_shortcode
+            performer.save()
+
+        model_run = ModelRun.objects.create(
+            title=model_run_upload.title,
+            performer=performer,
+            region=region,
+            parameters={},
+            public=not model_run_upload.private,
+        )
+
+        for site_model in site_models:
+            SiteEvaluation.bulk_create_from_site_model(site_model, model_run)
+
+        SiteEvaluation.bulk_create_from_region_model(region_model, model_run)
+
+
+@shared_task(bind=True)
+def process_model_run_upload_task(task, upload_id: UUID):
+    model_run_upload = ModelRunUpload.objects.get(pk=upload_id)
+
+    try:
+        with transaction.atomic():
+            model_run_upload.task_id = task.request.id
+            model_run_upload.save()
+
+        process_model_run_upload(model_run_upload)
+    finally:
+        model_run_upload.delete()
```

Comment:
I would pessimistically lock this if you are using .save(), since if the `model_run_upload` object becomes stale between the `.get` and `.save` call, the UPDATE query generated from .save() can clobber the row. 

```suggestion
    try:
        with transaction.atomic():
            model_run_upload = ModelRunUpload.objects.select_for_update().get(pk=upload_id)
            model_run_upload.task_id = task.request.id
            model_run_upload.save()

        process_model_run_upload(model_run_upload)
    finally:
        model_run_upload.delete()
```

Or better yet, just reduce it to a single UPDATE query and avoid the staleness question entirely
```suggestion
    ModelRunUpload.objects.filter(pk=upload_id).update(task_id=task.request.id)

    try:
        process_model_run_upload(ModelRunUpload.objects.get(pk=upload_id))
    finally:
        model_run_upload.delete()
```

Reply:
Agreed! Thanks for catching that. The single update query is what I'm looking to do.